storage/innobase/buf/buf0lru.cc

```diff
@@ -406,17 +406,32 @@ buf_block_t *buf_LRU_get_free_block(bool have_mutex)
     ut_ad(LRU_size <= BUF_LRU_MIN_LEN ||
           available >= scan_depth || buf_pool.need_LRU_eviction());
 
+    ut_d(bool signalled = false);
+
     if (UNIV_UNLIKELY(available < scan_depth) && LRU_size > BUF_LRU_MIN_LEN)
     {
       mysql_mutex_lock(&buf_pool.flush_list_mutex);
       if (!buf_pool.page_cleaner_active())
+      {
         buf_pool.page_cleaner_wakeup(true);
+        ut_d(signalled = true);
+      }
       mysql_mutex_unlock(&buf_pool.flush_list_mutex);
     }
 
     if (!have_mutex)
       mysql_mutex_unlock(&buf_pool.mutex);
 
+    DBUG_EXECUTE_IF("ib_free_page_sleep",
+    {
+      static bool do_sleep = true;
+      if (do_sleep && signalled)
+      {
+        std::this_thread::sleep_for(std::chrono::milliseconds(10));
+        do_sleep = false;
+      }
+    });
+
```

Comment:
This is debug injection code for one particular test, so this does not matter much. Nevertheless, I can’t resist some nitpicking:
* We could use `buf_pool.mutex` to protect the shared variable `do_sleep`. However,  in the `!have_mutex` case we would already have released that mutex.
* It would be more efficient to let global or `static` variables to be zero-initialized, so that no special initialization code has to be emitted. I would suggest `static bool skip_sleep;`.
* It would be more efficient to check the local variable `signalled` before loading the global variable from memory.
```c++
static bool skip_sleep;
if (signalled && !skip_sleep)
{
  skip_sleep= true;
  if (!have_mutex)
    mysql_mutex_unlock(&buf_pool.mutex);
  std::this_thread::sleep_for(std::chrono::milliseconds(10));
  if (!have_mutex)
    mysql_mutex_lock(&buf_pool.mutex);
}
```

Reply:
I see your point but we are supposed to use "ib_free_page_sleep" from one connection. Even if we need to set it across multiple connections, we would probably need to use thread local to keep the decision independent. Doing all that seems a bit of over engineering and I would rather like to skip it for future if any such need arises.